rust-symcrypt/src/ecdh.rs

```diff
@@ -0,0 +1,352 @@
+/* EcDh functions. For further documentation please refer to symcrypt.h */
+
+use crate::curve_type::*;
+use crate::errors::SymCryptError;
+use std::vec;
+use symcrypt_sys;
+
+pub struct EcDh {
+    curve_type: CurveType,
+    expanded_curve: Option<symcrypt_sys::PSYMCRYPT_ECURVE>,
+    key: Option<symcrypt_sys::PSYMCRYPT_ECKEY>,
+}
+
+impl EcDh {
+    /// EcDh::new() returns a EcDh struct that has a private/public key pair.
+    pub fn new(curve: CurveType) -> Result<Self, SymCryptError> {
+        let mut instance = EcDh {
+            curve_type: curve,
+            expanded_curve: None,
+            key: None,
+        };
+
+        unsafe {
+            // SAFETY: FFI calls
+            let curve_ptr =
+                symcrypt_sys::SymCryptEcurveAllocate(convert_curve(instance.curve_type), 0);
+            if curve_ptr.is_null() {
+                return Err(SymCryptError::MemoryAllocationFailure);
+            }
+
+            instance.expanded_curve = Some(curve_ptr);
+
+            let key_ptr = symcrypt_sys::SymCryptEckeyAllocate(curve_ptr);
+            if key_ptr.is_null() {
+                return Err(SymCryptError::MemoryAllocationFailure);
+            }
+
+            instance.key = Some(key_ptr);
+
+            symcrypt_sys::SymCryptEckeySetRandom(
```

Comment:
>symcrypt_sys::SymCryptEckeySetRandom [](http://example.com/codeflow?start=12&length=36)

This API can fail with SYMCRYPT_ERROR which needs to be returned. Can't panic for an error.

Reply:
Do same as you are doing in from_public_key_bytes